### DISCUSSION TITLE
Add a workaround for GTK 4 shadows

### DISCUSSION THREAD

**Discussion by SkyyySi** [yshui/picom#1221] [Configuraiton]
The currently included workaround for GTK 3 does not appear to work for GTK 4 based applications.

```
shadow-exclude = [
  "_GTK_FRAME_EXTENTS@:c"
];
```

This will hide tell picom not to draw shadows for GTK 3 shadows (which would otherwise be seen as a part of the window). It would be great if there were a way to activate the same for GTK 4.

**Comment by mpnordland** [comment]
Hi @SkyyySi,
I was looking into this issue as well and I found a Reddit post with a solution:
https://www.reddit.com/r/linux/comments/r91aqy/libawaita_makes_programs_look_terrible/hnapp2r/

Basically just add `"override_redirect = true"` to your `shadow-exclude`.

**Comment by diogofd8** [comment]
@mpnordland 
How did you fix the issue? I added "override_redirect = true" too the picom.conf but to no avail.
Then I tried putting the line inside "shadow-exclude" but couldn't find a way to make it accept the syntax. The reddit comment doesn't give away a solution unfortunately.

**Comment by vladvasiliu** [comment]
This worked for me:

```
shadow-exclude = [
    <bunch-of-rules>,
    "override_redirect = true",
];
```

**Comment by jEsuSdA** [comment]
@vladvasiliu 

Thanks a lot!

This solves my issue too:
https://github.com/yshui/picom/issues/806

Thanks!

---

**EDIT: Sorry! I just noticed this workaround disables ALL dropdown and menu shadows.
I think this is not the solution. :S**

**Comment by andregalastri** [comment]
I had a similar problem with the context menu of some GTK "Save as..." windows. I don't know if it would work for you, but here it is.

This was the issue I was facing:
![image](https://user-images.githubusercontent.com/49572917/184049521-71ea5f94-d4cf-4e44-9cf6-80588c68e1e7.png)

I solved it by including the following command in `picom.conf`
```
shadow-exclude = [
    <your-rules>,
    "class_g = 'xdg-desktop-portal-gnome' && window_type = 'menu'",
];
```
This will remove only shadows from the context menu. If you want to remove all shadows, just remove the `&& window_type = 'menu'` part, like this
```
shadow-exclude = [
    <your-rules>,
    "class_g = 'xdg-desktop-portal-gnome'",
];
```

Hope it helps

**Comment by Flashwalker** [comment]
> Hi @SkyyySi, I was looking into this issue as well and I found a Reddit post with a solution: https://www.reddit.com/r/linux/comments/r91aqy/libawaita_makes_programs_look_terrible/hnapp2r/
> 
> Basically just add `"override_redirect = true"` to your `shadow-exclude`.

It disables more then needed.

**Comment by Flashwalker** [comment]
i'm using this (related to GTK4 xdg-portal dialog and Firefox menus):

```
shadow-exclude = [
  ,"class_g ~= 'xdg-desktop-portal' && _NET_FRAME_EXTENTS@:c && window_type = 'dialog'"
  ,"class_g ~= 'xdg-desktop-portal' && window_type = 'menu'"
   "_NET_FRAME_EXTENTS@:c && WM_WINDOW_ROLE@:s = 'Popup'"
  ,"class_i = 'Firefox' && window_type = 'utility'"
  ,"class_i = 'Firefox' && window_type = 'popup_menu'"
];
blur-background-exclude = [
  ,"class_g ~= 'xdg-desktop-portal' && _NET_FRAME_EXTENTS@:c && window_type = 'dialog'"
  ,"class_g ~= 'xdg-desktop-portal' && window_type = 'menu'"
  ,"_NET_FRAME_EXTENTS@:c && WM_WINDOW_ROLE@:s = 'Popup'"
  ,"class_i = 'Firefox' && window_type = 'utility'"
  ,"class_i = 'Firefox' && window_type = 'popup_menu'"
];
```

**Comment by mgher668** [comment]
I solved this by adding menu rule to wintypes for menus/popovers in `picom.conf` file.

```
wintypes:
{
  menu = { shadow = false; blur-background = false; };
};
```

**Reply by brainwo** [reply]
This solution is perfect

**Comment by nestor-andreu** [comment]
I'd like to point out that this issue still persists. The idea is not to remove shadows or blur on all menus, but just for gtk popovers (not for Qt's, for example). Ideally, a workaround should fix *all* of gtk4 menus and popovers, not just a specific application (such as using `"class_g ~= 'xdg-desktop-portal'...`). I haven't been able to pinpoint the window properties of popovers, as xprop does not work when trying to click on one (for example, executing `sleep 5 && xprop`, then immediately opening a popover, xprop exits with a message `xprop: error: Can't grab the mouse`). So no luck yet, it seems. I'm on Manjaro (Arch derivative) and `xprop --version` outputs `vgit-89c2c` (and package version is 1.2.7-1). No luck so far. Has there been any step towards a solution or workaround? Thanks!

**Reply by Flashwalker** [reply]
I don't think this will be fixed by the picom developers as well as the gtk developers.

**Reply by 0pLuS0** [reply]
I would assume because the life of X is coming to end, unless picom can somehow convert over into something under Wayland next.

**Comment by starquake** [comment]
I found another workaround that seemed to work: (I don't know if there are any unwanted side-effects)

- Add `"argb && (override_redirect || wmwin)",` to `blur-background-exclude` and `shadow-exclude` [source](https://www.linuxquestions.org/questions/blog/the-dsc-472367/fix-google-chromes-ugly-thick-transparent-but-shadowed-borders-on-compositors-such-as-picom-39172/)

As described in the source above a way to use `xprop` is to start it, then focus the window using the keyboard, and opening the menu with the keyboard.

EDIT: Seems blur is missing from my semi-transparent terminal windows. Another try: `"name = '' && argb && (override_redirect || wmwin)"`

**Comment by 0pLuS0** [comment]
I have this issue with handbrake, KeePassXC and kid3-qt.

```"argb && (override_redirect || wmwin)", to blur-background-exclude and shadow-exclude``` seems to work.

But applications can be affected, so exclusions for them will need to be added.

Using this rofi will not have a shadow around it, this seems to correct, and allow it to have a shadow around it;

```"argb && (override_redirect || wmwin) && class_g != 'Rofi' && class_g != 'rofi'"```

I believe this is the correct way to exclude an app...

THANKS
